### ISSUE TITLE
Using Hookstate with memo?

### ISSUE THREAD

**Issue by mohit1234** [avkonst/hookstate#130]
First of all, this library is really amazing and simple to use. Really appreciate all your hard work on this! 😃 

Is it possible to use this library with memo? (I wasn't able to find any documentation)

I am trying to render a list where each item is very expensive to re-render. Anytime a new item is added to the list, the entire list rerenders and results in a noticeable lag. I was hoping to use memo on the list items to prevent that rerender but I'm not entirely sure how to get that work with hookstate. If I use the 2nd argument of React.memo to do a custom equality check, then I see that prevProps and nextProps have the same values for the properties that are hookstate states.

**Comment by avkonst** [comment]
Hi, could you please share with me a simple code sample of your component which you try to wrap in memo?

**Comment by mohit1234** [comment]
Thanks for the prompt reply and apologies for the delay. While I can't share the actual code, here's a similar sample:

```js
const MyComponent = ({ data, errorLogic, ...otherProps }) => {
    const dataState = useState(data);

    return (
        <>
            <MyList data={data} />
            <ErrorMessage ...otherProps errorLogic={errorLogic} />
        </>
    );
}

const MyList = ({ data }) => {
    return data.map(x => (<Collapsible unmountOnCollapse={true}><ListItem key={Untracked(x.id).get()} item={x} /></Collapsible>));
}

const ListItem = ({ item }) => {
    return (
        <>
            <ComponentA prop={item.propA} />
            <ComponentB prop={item.propB} />
            <VeryExpensiveComponent prop={item.propC} />
        </>
    );
}
```

So basically I have a component with a list of items. Each item in that list is collapsible and the item has one very expensive component. So when I collapse and uncollapse the item, it takes 2-3 seconds and the page hangs while the component is being rendered and added back to the dom.

My goal was to wrap the "ListItem" component in memo and use that in MyList so that there wouldn't be a need to rerender when the item is unmounted and remounted during collapse. Similarly, when an item is added or removed from the list, the entire tree might not be rerender (because of the key property) but React does try to do reconciliation which also takes time with this expensive component that I would like to avoid.

So if I wanted to wrap ListItem in memo, would that be possible given that some of the properties are hookstate objects?

**Comment by avkonst** [comment]
Do you realise that when you update any of the items (I assume collapse/uncollapse is a state update per item), the entire `MyComponent`, and so `MyList`, is rerendered? You would need to use scoped state to address it: https://hookstate.js.org/docs/scoped-state 
Using memo here would not help as much until you modify your example to use scoped state. If you prefer to go without scoped state and rely on memo dependency comparison (ala traditional React.useState + React.memo combo), you need to pass `Downgraded` state value to the memo (https://hookstate.js.org/docs/performance-managed-rendering/#downgraded-plugin) and come up with some mechanism of checking for state value equality, which needs to be supplied as the next argument to `memo` function after dependencies list.

**Comment by mohit1234** [comment]
All components are using scoped states so I don't think that's a concern. Also for collapse/uncollapse, its not based on a property in the item, its a local state in that component (not related to hookstate). So:
1. When an item is collapsed or uncollapsed, the entire list is not rerendered, only the item is. Because of the way collapse is setup, it will unmount the item when collapsed and remount on uncollapse. So here I was hoping to avoid rerender of the item by using memo
2. When an item is updated, only the related component is updated due to using nested and scope states so thats not an issue.

When an item is added/updated in the list, all items are rerendered and react does reconciliation which takes time which is what i'm trying to avoid.

Knowing all of this, would you still suggest using `Downgraded` + useMemo?

**Comment by avkonst** [comment]
Could you please update your example to show where you use the scoped state in the ListItem?

**Comment by avkonst** [comment]
Related https://github.com/avkonst/hookstate/issues/140

**Comment by avkonst** [comment]
It will be addressed in #140
